ghcide/src/Development/IDE/Spans/Common.hs

```diff
@@ -77,12 +79,19 @@ emptySpanDoc :: SpanDoc
 emptySpanDoc = SpanDocText [] (SpanDocUris Nothing Nothing)
 
 spanDocToMarkdown :: SpanDoc -> [T.Text]
-spanDocToMarkdown (SpanDocString docs uris)
-  = [T.pack $ haddockToMarkdown $ H.toRegular $ H._doc $ H.parseParas Nothing $ unpackHDS docs]
-    <> ["\n"] <> spanDocUrisToMarkdown uris
-  -- Append the extra newlines since this is markdown --- to get a visible newline,
-  -- you need to have two newlines
-spanDocToMarkdown (SpanDocText txt uris) = txt <> ["\n"] <> spanDocUrisToMarkdown uris
+spanDocToMarkdown = \case
+    (SpanDocString docs uris) ->
+        let doc = T.pack $ haddockToMarkdown $ H.toRegular $ H._doc $ H.parseParas Nothing $ unpackHDS docs
+        in  go [doc] uris
+    (SpanDocText txt uris) -> go txt uris
+  where
+    go [] uris = spanDocUrisToMarkdown uris
+    go txt uris = init txt <> [render (last txt)] <> spanDocUrisToMarkdown uris
+    -- If the doc is not end with an `'\n'`, we append it.
```

Comment:
This logic is a bit confusing, because we have a collection of lines, some of them ending with newlines. The old comment *is* correct, we do need two newlines to get a paragraph break in markdown, but that works out here because `T.unlines [ "foo\n", "bar" ]` will end up with two newlines. Confusing. 

In fact, should the lines _ever_ have trailing newlines? That seems odd.

I wonder if it would be simpler if this function just dealt with a single `Text` representing the whole piece of doc. AFAICT it would be fine if `spanDocToMarkdown` returned just a `Text`.

Reply:
1. Two newlines to get a new paragraph is correct, but one `\n` is enough to get a section separator(xxx\n***\nxxx)
2. My idea is: we just ensure the last line should end up with `\n`, so the separator can rendered correctly. https://github.com/haskell/haskell-language-server/blob/da0379b78caff2a3bc72e08b97b9f608fe73d50e/ghcide/src/Development/IDE/Plugin/Completions/Logic.hs#L211-L213

Reply:
>I wonder if it would be simpler if this function just dealt with a single Text representing the whole piece of doc. AFAICT it would be fine if spanDocToMarkdown returned just a Text.

Unfortunately, `spanDocToMarkdown` is both used in completion and hover, and they have different UI(completion uses section separator, hover uses new line), so we can't return a `Text` in `spanDocToMarkdown`. Furthermore, the final documentation is the combination of `Defined in`, haddock, and source link.

Reply:
Does it return a list of sections then? Please add a Haddock comment clarifying what `spanDocToMarkdown` returns

Reply:
Done